File: Adramelech/Commands/Slash/Anime.cs
using System.Diagnostics.CodeAnalysis;
using System.Text;
using Adramelech.Configuration;
using Adramelech.Extensions;
using Adramelech.Utilities;
using Discord;
using Discord.Interactions;
using Discord.WebSocket;
using Flurl;

namespace Adramelech.Commands.Slash;

[Group("anime", "Anime commands")]
public class Anime : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
{
    [Group("media", "Anime media commands")]
    public class Media(Config config, HttpUtils httpUtils)
        : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
    {
        [SlashCommand("image", "Get a random anime image")]
        public async Task ImageAsync(
            [Summary("age-rating", "The age rating of the image")] [Choice("SFW", "sfw")] [Choice("NSFW", "nsfw")]
            string ageRating = "sfw")
        {
            await DeferAsync();

            string[] rating;

            switch (ageRating)
            {
                case "sfw":
                    rating = ["safe", "questionable"];
                    break;
                case "nsfw":
                    if (Context.Channel is not SocketTextChannel { IsNsfw: true })
                    {
                        await Context.SendError("This command can only be used in NSFW channels.", true);
                        return;
                    }

                    // rating = ["borderline", "explicit"];
                    rating = ["explicit"];
                    break;
                default:
                    await Context.SendError("Invalid age rating.", true);
                    return;
            }

            var url = new Url("https://api.nekosapi.com")
                .AppendPathSegments("v3", "images", "random")
                .SetQueryParam("limit", 1)
                .SetQueryParam("rating", rating)
                .ToString()!;

            var response = await httpUtils.GetAsync<NekosApiResponse>(url, Config.UserAgent);
            if (response.IsDefault())
            {
                await Context.SendError("Failed to fetch image.", true);
                return;
            }

            var data = response.Items.First();

            var footer = new StringBuilder();
            if (data.Source is not null)
                footer.AppendLine($"Source: {data.Source}");
            footer.Append("Powered by nekosapi.com");

            await FollowupAsync(embed: new EmbedBuilder()
                .WithColor(config.EmbedColor)
                .WithImageUrl(data.ImageUrl)
                .WithFooter(footer.ToString())
                .Build());
        }

        [SlashCommand("neko", "Get a random neko image")]
        public async Task NekoAsync()
        {
            await DeferAsync();

            var response = await httpUtils.GetAsync<NekosLifeResponse>("https://nekos.life/api/v2/img/neko");
            if (response.IsDefault())
            {
                await Context.SendError("Failed to fetch image.", true);
                return;
            }

            await FollowupAsync(embed: new EmbedBuilder()
                .WithColor(config.EmbedColor)
                .WithImageUrl(response.Url)
                .WithFooter("Powered by nekos.life")
                .Build());
        }

        [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
        private struct NekosApiResponse
        {
            public SItem[] Items { get; set; }

            internal struct SItem
            {
                public string ImageUrl { get; set; }
                public string? Source { get; set; }
            }
        }

        [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
        private struct NekosLifeResponse
        {
            public string Url { get; set; }
        }
    }
}

File: bot/src/commands/slash/anime.ts
import {
  ComponentType,
  MessageFlags,
  SlashCommandBuilder,
  type ChatInputCommandInteraction,
  type TextChannel,
} from 'discord.js';
import ky from 'ky';
import {
  executeCommandFromTree,
  type CommandGroupExecutors,
  type CommandInfer,
} from '~/types/command';
import { sendError } from '~/utils/sendError';
import config from '~/config';
import StringBuilder from '~/tools/StringBuilder';
import { fromAsyncThrowable } from 'neverthrow';
import { type } from 'arktype';
import { capitalize } from 'utils/text';
import { arkToResult } from 'utils/validation';

const AnimeImageRatings = [
  'safe',
  'suggestive',
  'borderline',
  'explicit',
] as const;
const AnimeImageAgeRating = type
  .enumerated(...AnimeImageRatings)
  .or('null')
  .pipe((v) => v || 'safe');

const AnimeImages = type({
  url: 'string.url',
  source_url: 'string.url | null',
})
  .array()
  .exactlyLength(1);

const NekoImage = type({
  url: 'string.url',
});

export const command = <CommandInfer>{
  data: new SlashCommandBuilder()
    .setName('anime')
    .setDescription('Anime related commands')
    .addSubcommandGroup((group) =>
      group
        .setName('media')
        .setDescription('Anime media commands')
        .addSubcommand((subcommand) =>
          subcommand
            .setName('image')
            .setDescription('Get a random anime image')
            .addStringOption((option) =>
              option
                .setName('rating')
                .setDescription('The rating of the image')
                .setChoices(
                  AnimeImageRatings.map((rating) => ({
                    name: capitalize(rating),
                    value: rating,
                  })),
                ),
            ),
        )
        .addSubcommand((subcommand) =>
          subcommand.setName('neko').setDescription('Get a random neko image'),
        ),
    ),
  cooldown: true,
  uses: ['nekosapi.com', 'nekos.life'],
  execute: async (intr: ChatInputCommandInteraction) =>
    executeCommandFromTree(executors, intr),
};

const executors: CommandGroupExecutors = {
  media: {
    image: animeImage,
    neko: nekoImage,
  },
};

async function animeImage(intr: ChatInputCommandInteraction) {
  await intr.deferReply();

  const rating = AnimeImageAgeRating(intr.options.getString('rating'));
  if (rating instanceof type.errors)
    return await sendError(intr, 'Invalid rating');

  if (
    !(intr.channel as TextChannel).nsfw &&
    ['borderline', 'explicit'].includes(rating)
  )
    return await sendError(
      intr,
      'This command can only be used in NSFW channels',
    );

  const result = await fromAsyncThrowable(
    ky.get('https://api.nekosapi.com/v4/images/random', {
      searchParams: {
        rating,
        limit: 1,
      },
      headers: {
        'User-Agent': config.USER_AGENT,
      },
    }).json,
    (e) => `Failed to fetch image:\n${String(e)}`,
  )().andThen(arkToResult(AnimeImages));
  if (result.isErr()) return await sendError(intr, result.error);

  const footer = new StringBuilder();
  if (result.value[0]?.source_url)
    footer.appendLine(`> Source: ${result.value[0].source_url}`);
  footer.append(`> Powered by NekosAPI`);

  await intr.followUp({
    flags: MessageFlags.IsComponentsV2,
    components: [
      {
        type: ComponentType.Container,
        accent_color: config.EMBED_COLOR,
        components: [
          {
            type: ComponentType.MediaGallery,
            items: [
              {
                media: {
                  url: result.value[0]!.url,
                },
              },
            ],
          },
          {
            type: ComponentType.TextDisplay,
            content: footer.toString(),
          },
        ],
      },
    ],
  });
}

async function nekoImage(intr: ChatInputCommandInteraction) {
  await intr.deferReply();

  const result = await fromAsyncThrowable(
    ky.get('https://nekos.life/api/v2/img/neko').json,
    (e) => `Failed to fetch image:\n${String(e)}`,
  )().andThen(arkToResult(NekoImage));
  if (result.isErr()) return await sendError(intr, result.error);

  await intr.followUp({
    flags: MessageFlags.IsComponentsV2,
    components: [
      {
        type: ComponentType.Container,
        accent_color: config.EMBED_COLOR,
        components: [
          {
            type: ComponentType.MediaGallery,
            items: [
              {
                media: {
                  url: result.value.url,
                },
              },
            ],
          },
          {
            type: ComponentType.TextDisplay,
            content: '> Powered by nekos.life',
          },
        ],
      },
    ],
  });
}
